Update AssemblyBuilder.md with persistable AssemblyBuilder info (#39491)
dotnet/docs: docs/fundamentals/runtime-libraries/system-reflection-emit-assemblybuilder.md

[!INCLUDE [context](includes/context.md)]

> A dynamic assembly is an assembly that is created using the Reflection Emit APIs. A dynamic assembly can reference types defined in another dynamic or static assembly. You can use <xref:System.Reflection.Emit.AssemblyBuilder> to generate dynamic assemblies in memory and execute their code during the same application run. In .NET 9.0 we added new fully managed implementation for reflection emit that allows you save the asembly into a file. In .NET Framework, you can do both, run the dynamic assembly and save the assemblies to files. The dynamic assembly created for saving is called a *persistable* assembly, while the regular memory-only assembly is called *transient*. In .NET Framework, a dynamic assembly can consist of one or more dynamic modules. In .NET Core and .NET 5+, a dynamic assembly can only consist of one dynamic module.

Reviewer: ```suggestion
A dynamic assembly is an assembly that is created using the Reflection Emit APIs. A dynamic assembly can reference types defined in another dynamic or static assembly. You can use <xref:System.Reflection.Emit.AssemblyBuilder> to generate dynamic assemblies in memory and execute their code during the same application run. .NET 9 added a fully managed implementation for reflection emit that allows you save the assembly into a file. In .NET Framework, you can do both&mdash;run the dynamic assembly and save it to a file. The dynamic assembly created for saving is called a *persistable* assembly, while the regular memory-only assembly is called *transient*. In .NET Framework, a dynamic assembly can consist of one or more dynamic modules. In .NET Core and .NET 5+, a dynamic assembly can only consist of one dynamic module.
```

> The way of creating <xref:System.Reflection.Emit.AssemblyBuilder> instance differs for each implementation, but further steps for defining a module/type/method/enum etc., and writing IL are quite similar:

Reviewer: ```suggestion
The way you create an <xref:System.Reflection.Emit.AssemblyBuilder> instance differs for each implementation, but further steps for defining a module, type, method, or enum, and for writing IL, are quite similar.
```

The access mode must be specified by providing the appropriate <xref:System.Reflection.Emit.AssemblyBuilderAccess> value in the call to the <xref:System.Reflection.Emit.AssemblyBuilder.DefineDynamicAssembly%2A?displayProperty=nameWithType> method when the dynamic assembly is defined and cannot be changed later. The runtime uses the access mode of a dynamic assembly to optimize the assembly's internal representation.

> The following example demonstrates how to create and run assembly:

Reviewer: ```suggestion
The following example demonstrates how to create and run an assembly:
```

> ```cs

Reviewer: ```suggestion
```csharp
```

public void CreateAndRunAssembly(string assemblyPath)
{
    AssemblyBuilder ab = AssemblyBuilder.DefineDynamicAssembly(new AssemblyName("MyAssembly"), AssemblyBuilderAccess.Run);
    ModuleBuilder mob = ab.DefineDynamicModule("MyModule");
    TypeBuilder tb = mob.DefineType("MyType", TypeAttributes.Public | TypeAttributes.Class);
    MethodBuilder mb = tb.DefineMethod("SumMethod", MethodAttributes.Public | MethodAttributes.Static,
                                                                   typeof(int), new Type[] {typeof(int), typeof(int)});
    ILGenerator il = mb.GetILGenerator();
    il.Emit(OpCodes.Ldarg_0);
    il.Emit(OpCodes.Ldarg_1);
    il.Emit(OpCodes.Add);
    il.Emit(OpCodes.Ret);

    Type type = tb.CreateType();
    
    MethodInfo method = type.GetMethod("SumMethod");
    Console.WriteLine(method.Invoke(null, new object[] { 5, 10 }));
}
```

## Persistable dynamic assemblies in .NET Core

> The `AssemblyBuilder.Save` API was not ported to .NET core because of technical difficulties like the implementation heavily depend on windows specific native code that also not ported into .NET Core. When .NET Core developers can only run the generated assemblies it was very difficult to debug in-memory assemblies without having an option to save an assembly. Saving the assembly to a file allows generated assemblies to be verified with tools such as ILVerify, or decompiled and manually examined with tools such as ILSpy. Furthermore, the saved assembly can be shared or loaded directly which can be used to decrease application startup time.

Reviewer: ```suggestion
The `AssemblyBuilder.Save` API was not ported to .NET (Core) because the implementation depended heavily on Windows-specific native code that also wasn't ported. Because you could only *run* a generated assembly and not *save* it, it was difficult to debug these in-memory assemblies. Saving the assembly to a file lets you verify the generated assembly with tools such as ILVerify, or decompile and manually examine it with tools such as ILSpy. Furthermore, the saved assembly can be shared or loaded directly, which can decrease application startup time.
```

> In .NET 9.0 we are adding fully managed new `Reflection.Emit` implementation that supports saving. This implementation has no dependency from existing runtime specific `Reflection.Emit` implementation, i.e. now we have two different implementations in .NET Core. The assemblies generated with new managed implementation can be saved, to run the assembly user should save it into a memory stream or a file first, then load it back.  

Reviewer: ```suggestion
.NET 9 adds a fully managed `Reflection.Emit` implementation that supports saving. This implementation has no dependency on the pre-existing, runtime-specific `Reflection.Emit` implementation. That is, now there are two different implementations in .NET (Core). The assemblies generated with the new managed implementation can be saved. To run the assembly, first save it into a memory stream or a file, then load it back.  
```

> For creating a new persistable `AssemblyBuilder` instance you should use the new static factory method `AssemblyBuilder.DefinePersistedAssembly`:

Reviewer: ```suggestion
To create a new persistable `AssemblyBuilder` instance, use the static factory method `AssemblyBuilder.DefinePersistedAssembly`:
```

```cs
public static DefinePersistedAssembly(AssemblyName name, Assembly coreAssembly,
                                             IEnumerable<CustomAttributeBuilder>? assemblyAttributes = null);
```

> The `coreAssembly` passed to the method used for resolving base runtime types and can be used for resolving reference assemblies versioning.

Reviewer: ```suggestion
The `coreAssembly` parameter is used to resolve base runtime types and can be used for resolving reference assembly versioning.
```

```cs
public void CreateSaveAndRunAssembly(string assemblyPath)
{
    AssemblyBuilder ab = AssemblyBuilder.DefinePersistedAssembly(new AssemblyName("MyAssembly"), typeof(object).Assembly);
    ModuleBuilder mob = ab.DefineDynamicModule("MyModule");
    TypeBuilder tb = mob.DefineType("MyType", TypeAttributes.Public | TypeAttributes.Class);
    MethodBuilder meb = tb.DefineMethod("SumMethod", MethodAttributes.Public | MethodAttributes.Static,
                                                                   typeof(int), new Type[] {typeof(int), typeof(int)});
    ILGenerator il = meb.GetILGenerator();
    il.Emit(OpCodes.Ldarg_0);
    il.Emit(OpCodes.Ldarg_1);
    il.Emit(OpCodes.Add);
    il.Emit(OpCodes.Ret);

    tb.CreateType();

    using var stream = new MemoryStream();
    ab.Save(stream);  // or pass filename to save into a file
    stream.Seek(0, SeekOrigin.Begin);
    Assembly assembly = AssemblyLoadContext.Default.LoadFromStream(stream);
    MethodInfo method = assembly.GetType("MyType").GetMethod("SumMethod");
    Console.WriteLine(method.Invoke(null, new object[] { 5, 10 }));
}
```

> [!NOTE]

> > The metadata tokens for all members are populated on `Save(...)` operation, do not use the tokens of generated type and its members before saving as they will have default values or throw. It is safe to use tokens for types that are referenced, not generated.
> > Some APIs that are not improtant for emitting assembly are not implemented, for example `GetCustomAttributes()` is not implemented, with the runtime implementation you were able to use those APIs after creating the type, for persisted `AssemblyBuilder` they would throw `NotSupportedException` or `NotImplementedException`. If you have a scenario that needs those APIs implemented file an issue in the [repo](https://github.com/dotnet/runtime).

Author: ```suggestion
> * The metadata tokens for all members are populated on `Save(...)` operation, do not use the tokens of generated type and its members before saving as they will have default values or throw. It is safe to use tokens for types that are referenced, not generated.
>
> * Some APIs that are not improtant for emitting assembly are not implemented, for example `GetCustomAttributes()` is not implemented, with the runtime implementation you were able to use those APIs after creating the type, for persisted `AssemblyBuilder` they would throw `NotSupportedException` or `NotImplementedException`. If you have a scenario that needs those APIs implemented file an issue in the [repo](https://github.com/dotnet/runtime).
```

}
```

## Persistable dynamic assemblies in .NET Core

The `AssemblyBuilder.Save` API was not ported to .NET (Core) because the implementation depended heavily on Windows-specific native code that also wasn't ported. Because you could only *run* a generated assembly and not *save* it, it was difficult to debug these in-memory assemblies. Saving the assembly to a file lets you verify the generated assembly with tools such as ILVerify, or decompile and manually examine it with tools such as ILSpy. Furthermore, the saved assembly can be shared or loaded directly, which can decrease application startup time.

.NET 9 adds a fully managed `Reflection.Emit` implementation that supports saving. This implementation has no dependency on the pre-existing, runtime-specific `Reflection.Emit` implementation. That is, now there are two different implementations in .NET (Core). The assemblies generated with the new managed implementation can be saved. To run the assembly, first save it into a memory stream or a file, then load it back.  

To create a new persistable `AssemblyBuilder` instance, use the static factory method `AssemblyBuilder.DefinePersistedAssembly`:

> ```cs

Author: ```suggestion
```csharp
```

public static DefinePersistedAssembly(AssemblyName name, Assembly coreAssembly,
                                             IEnumerable<CustomAttributeBuilder>? assemblyAttributes = null);
```

The `coreAssembly` parameter is used to resolve base runtime types and can be used for resolving reference assembly versioning.

> - if `Reflection.Emit` is used to generate assembly that targets specific TFM, the reference assemblies for the given TFM should be opened using `MetadataLoadContext` and the value of [MetadataLoadContext.CoreAssembly](https://learn.microsoft.com/dotnet/api/system.reflection.metadataloadcontext.coreassembly) property should be used as the coreAssembly. It allows the generator to run on one .NET runtime version and target a different .NET runtime version.

Reviewer: ```suggestion
- If `Reflection.Emit` is used to generate an assembly that targets a specific TFM, open the reference assemblies for the given TFM using `MetadataLoadContext` and use the value of the [MetadataLoadContext.CoreAssembly](xref:System.Reflection.MetadataLoadContext.CoreAssembly) property for `coreAssembly`. This value allows the generator to run on one .NET runtime version and target a different .NET runtime version.
```

> - If `Reflection.Emit` is used to generate an assembly that is only going to be executed on the same runtime version as the runtime version that the compiler is running on (typically in-proc), the core assembly can be `typeof(object).Assembly`. The reference assemblies are not necessary in this case.

Reviewer: ```suggestion
- If `Reflection.Emit` is used to generate an assembly that's only going to be executed on the same runtime version as the runtime version that the compiler is running on (typically in-proc), the core assembly can be `typeof(object).Assembly`. The reference assemblies aren't necessary in this case.
```

The following example demonstrates how to create and save an assembly to a stream and run it:

> ```cs

Author: ```suggestion
```csharp
```

public void CreateSaveAndRunAssembly(string assemblyPath)
{
    AssemblyBuilder ab = AssemblyBuilder.DefinePersistedAssembly(new AssemblyName("MyAssembly"), typeof(object).Assembly);
    ModuleBuilder mob = ab.DefineDynamicModule("MyModule");
    TypeBuilder tb = mob.DefineType("MyType", TypeAttributes.Public | TypeAttributes.Class);
    MethodBuilder meb = tb.DefineMethod("SumMethod", MethodAttributes.Public | MethodAttributes.Static,
                                                                   typeof(int), new Type[] {typeof(int), typeof(int)});
    ILGenerator il = meb.GetILGenerator();
    il.Emit(OpCodes.Ldarg_0);
    il.Emit(OpCodes.Ldarg_1);
    il.Emit(OpCodes.Add);
    il.Emit(OpCodes.Ret);

    tb.CreateType();

    using var stream = new MemoryStream();
    ab.Save(stream);  // or pass filename to save into a file
    stream.Seek(0, SeekOrigin.Begin);
    Assembly assembly = AssemblyLoadContext.Default.LoadFromStream(stream);
    MethodInfo method = assembly.GetType("MyType").GetMethod("SumMethod");
    Console.WriteLine(method.Invoke(null, new object[] { 5, 10 }));
}
```

> [!NOTE]

> > The metadata tokens for all members are populated on `Save(...)` operation, do not use the tokens of generated type and its members before saving as they will have default values or throw. It is safe to use tokens for types that are referenced, not generated.

Reviewer: ```suggestion
> The metadata tokens for all members are populated on the `Save(...)` operation. Don't use the tokens of a generated type and its members before saving, as they'll have default values or throw exceptions. It's safe to use tokens for types that are referenced, not generated.
```

> > * The metadata tokens for all members are populated on `Save(...)` operation, do not use the tokens of generated type and its members before saving as they will have default values or throw. It is safe to use tokens for types that are referenced, not generated.
> >
> > * Some APIs that are not improtant for emitting assembly are not implemented, for example `GetCustomAttributes()` is not implemented, with the runtime implementation you were able to use those APIs after creating the type, for persisted `AssemblyBuilder` they would throw `NotSupportedException` or `NotImplementedException`. If you have a scenario that needs those APIs implemented file an issue in the [repo](https://github.com/dotnet/runtime).

Author: ```suggestion
> The metadata tokens for all members are populated on `Save(...)` operation, do not use the tokens of generated type and its members before saving as they will have default values or throw. It is safe to use tokens for types that are referenced, not generated.
>
> Some APIs that are not improtant for emitting assembly are not implemented, for example `GetCustomAttributes()` is not implemented, with the runtime implementation you were able to use those APIs after creating the type, for persisted `AssemblyBuilder` they would throw `NotSupportedException` or `NotImplementedException`. If you have a scenario that needs those APIs implemented file an issue in the [repo](https://github.com/dotnet/runtime).
```

> > Some APIs that are not improtant for emitting assembly are not implemented, for example `GetCustomAttributes()` is not implemented, with the runtime implementation you were able to use those APIs after creating the type, for persisted `AssemblyBuilder` they would throw `NotSupportedException` or `NotImplementedException`. If you have a scenario that needs those APIs implemented file an issue in the [repo](https://github.com/dotnet/runtime).

Reviewer: ```suggestion
> Some APIs that aren't important for emitting an assembly aren't implemented; for example, `GetCustomAttributes()` is not implemented. With the runtime implementation, you were able to use those APIs after creating the type. For the persisted `AssemblyBuilder`, they throw `NotSupportedException` or `NotImplementedException`. If you have a scenario that requires those APIs, file an issue in the [dotnet/runtime repo](https://github.com/dotnet/runtime).
```

For an alternative way to generate assembly files, see <xref:System.Reflection.Metadata.Ecma335.MetadataBuilder>.

## Persistable dynamic assemblies in .NET Framework

In .NET Framework, dynamic assemblies and modules can be saved to files. To support this feature, the <xref:System.Reflection.Emit.AssemblyBuilderAccess> enumeration declares two additional fields: <xref:System.Reflection.Emit.AssemblyBuilderAccess.Save> and <xref:System.Reflection.Emit.AssemblyBuilderAccess.RunAndSave>.

The dynamic modules in the persistable dynamic assembly are saved when the dynamic assembly is saved using the <xref:System.Reflection.Emit.AssemblyBuilder.Save%2A> method. To generate an executable, the <xref:System.Reflection.Emit.AssemblyBuilder.SetEntryPoint%2A> method must be called to identify the method that is the entry point to the assembly. Assemblies are saved as DLLs by default, unless the <xref:System.Reflection.Emit.AssemblyBuilder.SetEntryPoint%2A> method requests the generation of a console application or a Windows-based application.

> The following example demonstrates how to create and save and run assembly using .NET Framework:

Reviewer: ```suggestion
The following example demonstrates how to create, save, and run an assembly using .NET Framework.
```

> ```cs

Author: ```suggestion
```csharp
```

public void CreateRunAndSaveAssembly(string assemblyPath)
{
    AssemblyBuilder ab = Thread.GetDomain().DefineDynamicAssembly(new AssemblyName("MyAssembly"), AssemblyBuilderAccess.RunAndSave);
    ModuleBuilder mob = ab.DefineDynamicModule("MyAssembly.dll");
    TypeBuilder tb = mob.DefineType("MyType", TypeAttributes.Public | TypeAttributes.Class);
    MethodBuilder meb = tb.DefineMethod("SumMethod", MethodAttributes.Public | MethodAttributes.Static,
                                                                   typeof(int), new Type[] {typeof(int), typeof(int)});
    ILGenerator il = meb.GetILGenerator();
    il.Emit(OpCodes.Ldarg_0);
    il.Emit(OpCodes.Ldarg_1);
    il.Emit(OpCodes.Add);
    il.Emit(OpCodes.Ret);

    Type type = tb.CreateType();

    MethodInfo method = type.GetMethod("SumMethod");
    Console.WriteLine(method.Invoke(null, new object[] { 5, 10 }));
    ab.Save("MyAssembly.dll");
}
```

> Some methods on the base <xref:System.Reflection.Assembly> class, such as `GetModules` and `GetLoadedModules`, will not work correctly when called from <xref:System.Reflection.Emit.AssemblyBuilder> objects. You can load the defined dynamic assembly and call the methods on the loaded assembly. For example, to ensure that resource modules are included in the returned module list, call `GetModules` on the loaded <xref:System.Reflection.Assembly> object. If a dynamic assembly contains more than one dynamic module, the assembly's manifest file name should match the module's name that is specified as the first argument to the <xref:System.Reflection.Emit.AssemblyBuilder.DefineDynamicModule%2A> method.

Reviewer: ```suggestion
Some methods on the base <xref:System.Reflection.Assembly> class, such as `GetModules` and `GetLoadedModules`, won't work correctly when called from <xref:System.Reflection.Emit.AssemblyBuilder> objects. You can load the defined dynamic assembly and call the methods on the loaded assembly. For example, to ensure that resource modules are included in the returned module list, call `GetModules` on the loaded <xref:System.Reflection.Assembly> object. If a dynamic assembly contains more than one dynamic module, the assembly's manifest file name should match the module's name that's specified as the first argument to the <xref:System.Reflection.Emit.AssemblyBuilder.DefineDynamicModule%2A> method.
```
